File: Android_Zone_Test/src/com/example/mylib_test/activity/animal/WaveActivity.java
package com.example.mylib_test.activity.animal;

import android.app.Activity;
import android.graphics.Bitmap;
import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.Matrix;
import android.graphics.PorterDuff;
import android.os.Bundle;
import android.widget.ImageView;
import android.widget.SeekBar;

import com.example.mylib_test.R;

import and.utils.image.WaveHelper;
import and.utils.image.compress2sample.SampleUtils;
import and.utils.image.BitmapComposer;
import view.WaveView2;

//参考:https://github.com/race604/WaveLoading
public class WaveActivity extends Activity {
    SeekBar mLevelSeekBar;
    SeekBar mAmplitudeSeekBar;
    SeekBar mLengthSeekBar;
    SeekBar mSpeedSeekBar;
    WaveView2 waveView;
    SeekBar speed_OffsetXLength;
    private ImageView image2;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.a_animal_wave);
        waveView = (WaveView2) findViewById(R.id.bt_wave);
        mLevelSeekBar = (SeekBar) findViewById(R.id.level_seek);
        mLevelSeekBar.setOnSeekBarChangeListener(new SimpleOnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                waveView.setLevel(((float) progress) / seekBar.getMax());
            }
        });

        mAmplitudeSeekBar = (SeekBar) findViewById(R.id.amplitude_seek);
        mAmplitudeSeekBar.setOnSeekBarChangeListener(new SimpleOnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                waveView.setWaveAmplitude(progress);
            }
        });

        mLengthSeekBar = (SeekBar) findViewById(R.id.length_seek);
        mLengthSeekBar.setOnSeekBarChangeListener(new SimpleOnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                waveView.setWaveLength(progress);
            }
        });

        mSpeedSeekBar = (SeekBar) findViewById(R.id.speed_seek);
        mSpeedSeekBar.setOnSeekBarChangeListener(new SimpleOnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                waveView.setWaveSpeed(progress);
            }
        });
        speed_OffsetXLength = (SeekBar) findViewById(R.id.speed_OffsetXLength);
        speed_OffsetXLength.setOnSeekBarChangeListener(new SimpleOnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                waveView.setOffsetXRadioOfLength(((float) progress) / seekBar.getMax());
            }
        });

        image2 = (ImageView) findViewById(R.id.image2);
        image2.post(new Runnable() {
            @Override
            public void run() {
                Bitmap bt = SampleUtils.load(WaveActivity.this, R.drawable.aaaaaaaaaaaab)
                        .bitmap();

                ColorMatrix colorMatrix = new ColorMatrix();
                colorMatrix.setSaturation(0);

                BitmapComposer bitmapComposer = BitmapComposer.newComposition(bt.getWidth(), bt.getHeight(), Bitmap.Config.ARGB_8888);
                Matrix first = new Matrix();
                first.postTranslate(0, -20);
                new WaveHelper(bt.getWidth(), bt.getHeight(), new WaveHelper.RefreshCallback() {
                    public BitmapComposer.Layer mb;
                    public BitmapComposer.Layer kb;

                    @Override
                    public void refresh(Bitmap wave) {
                        Bitmap render = bitmapComposer.clear()
                                .newLayer(kb=BitmapComposer.Layer.bitmap(bt)
                                        .colorFilter(new ColorMatrixColorFilter(colorMatrix)))
                                .newLayer(
                                        mb=BitmapComposer.Layer.bitmap(bt)
//                                                .colorFilter(new ColorMatrixColorFilter(colorMatrix))
                                                .mask(wave, PorterDuff.Mode.DST_IN)
//                                                .matrix(first)
                                )
                                .render();
                        image2.setImageBitmap(render);
                    }
                });

            }
        });


    }

    private static class SimpleOnSeekBarChangeListener implements SeekBar.OnSeekBarChangeListener {

        @Override
        public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            // Nothing
        }

        @Override
        public void onStartTrackingTouch(SeekBar seekBar) {
            // Nothing
        }

        @Override
        public void onStopTrackingTouch(SeekBar seekBar) {
            // Nothing
        }
    }

}

File: Android_Zone_Test/src/com/example/mylib_test/activity/animal/WaveActivity.kt
package com.example.mylib_test.activity.animal

import android.graphics.Bitmap
import android.graphics.ColorMatrix
import android.graphics.ColorMatrixColorFilter
import android.graphics.Matrix
import android.graphics.PorterDuff
import android.widget.SeekBar

import com.example.mylib_test.R
import com.zone.lib.base.controller.activity.BaseFeatureActivity

import com.zone.lib.utils.image.WaveHelper
import com.zone.lib.utils.image.compress2sample.SampleUtils
import com.zone.lib.utils.image.BitmapComposer
import kotlinx.android.synthetic.main.a_animal_wave.*

//参考:https://github.com/race604/WaveLoading
class WaveActivity : BaseFeatureActivity() {
    var waveHelper: WaveHelper? = null
    override fun setContentView() {
        setContentView(R.layout.a_animal_wave)
    }

    override fun initData() {
        image2.post {
            val bt = SampleUtils.load(this@WaveActivity, R.drawable.aaaaaaaaaaaab)
                    .bitmap()
            val colorMatrix = ColorMatrix()
            colorMatrix.setSaturation(0f)

            val bitmapComposer = BitmapComposer.newComposition(bt.width, bt.height, Bitmap.Config.ARGB_8888)
            val first = Matrix()
            first.postTranslate(0f, -20f)
            waveHelper = WaveHelper(bt.width, bt.height, object : WaveHelper.RefreshCallback {
                override fun refresh(wave: Bitmap) {
                    val mb = BitmapComposer.Layer.bitmap(bt)
                            .colorFilter(ColorMatrixColorFilter(colorMatrix))
                    val kb = BitmapComposer.Layer.bitmap(bt)
//    .colorFilter(new ColorMatrixColorFilter(colorMatrix))
                            .mask(wave, PorterDuff.Mode.DST_IN)
//                            .matrix(first)
                    val render = bitmapComposer.clear()
                            .newLayer(mb)
                            .newLayer(kb)
                            .render()
                    image2.setImageBitmap(render)
                }
            })
        }
    }

    override fun setListener() {
        level_seek.setOnSeekBarChangeListener(object : SimpleOnSeekBarChangeListener() {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                bt_wave.setLevel(progress.toFloat() / seekBar.max)
                waveHelper?.setLevelProgress(progress.toFloat() / seekBar.max)
            }
        })
        amplitude_seek.setOnSeekBarChangeListener(object : SimpleOnSeekBarChangeListener() {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                bt_wave.setWaveAmplitude(progress)
                waveHelper?.amplitude = progress.toFloat()
            }
        })

        length_seek.setOnSeekBarChangeListener(object : SimpleOnSeekBarChangeListener() {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                bt_wave.setWaveLength(progress)
                waveHelper?.length = progress.toFloat()
            }
        })

        speed_seek.setOnSeekBarChangeListener(object : SimpleOnSeekBarChangeListener() {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                bt_wave.setWaveSpeed(progress)
                waveHelper?.speed = progress.toFloat()
            }
        })
        speed_OffsetXLength.setOnSeekBarChangeListener(object : SimpleOnSeekBarChangeListener() {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                bt_wave.setOffsetXRadioOfLength(progress.toFloat() / seekBar.max)
                waveHelper?.setOffsetXRadioOfLength(progress.toFloat() / seekBar.max)
            }
        })


    }

    open class SimpleOnSeekBarChangeListener : SeekBar.OnSeekBarChangeListener {

        override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
            // Nothing
        }

        override fun onStartTrackingTouch(seekBar: SeekBar) {
            // Nothing
        }

        override fun onStopTrackingTouch(seekBar: SeekBar) {
            // Nothing
        }
    }

}
